simple core concept split structure (#20)
comit-network/comit.network: docs/core-concepts/atomic-swap-htlc.md

> ---

Author: 👍 I'll see what I can do, thanks for the structure proposal!

id: atomic-swap-htlc
title: Atomic Swaps using HTLCs
sidebar_label: Atomic Swaps using HTLCs
---

[Atomic Swap](#atomic-swap)s using [HTLC](#htlc)s defines a solution using timelocks and hashlocks for exchanging two digital assets without the need of a trusted third party.

Prior to the actual swap execution there is usually a [Negotiation](negotiation.md) phase where the two parties agree on the ledgers, assets and amounts to be swapped.

Assuming that Alice has ETH on Ethereum and Bob has BTC on Bitcoin, Alice wants Bob's BTC and Bob wants Alice's ETH, the execution of a swap would be like this:

> 1. Alice generates a secret and hashes it. She sends to Bob:

Reviewer: I am assuming you are describing atomic swaps with HTLCs _in general_ here and not how exactly we are implementing them.

Re: "She sends to Bob":
You are mixing our current implementation (which to some extend is an implementation detail) with the general protocol of atomic swaps.

If we don't want this this documentation to be outdated, I would omit these kind of details here and explain atomic swaps using HTLCs without caring about the communication (you could say something like "they agree on identities, amounts and expiry times").

Author: I can see why you would raise this, and I would like to have more opinions on that. To some extent I want to be objective in this section and describe concepts outside of COMIT, however, I also feel we can put *some* influence of our implementation details into the core concept documentation. This way it might be easier for users to understand our solution.

That might be a bit far fetched though 😅 
@bonomat @D4nte @yosriady what is your opinion on that? Should I change this to be more general? (this would mean, that I would add another sentence before the slide showing the swap, to explain that this is how we do it in COMIT...)

Reviewer 2: I tend to agree with @thomas here.  Keeping it more general makes it read more objective, which I think is a good thing to aim for.  Good work sticking at this @da-kami, writing good docs is tough.

    1. Hash of the secret. 
    1. Her refund address on Ethereum.
    1. Her redeem address on Bitcoin.

>     1. Proposed expiry times on the for the [HTLC](#htlc)s.

Reviewer 2: ```suggestion
    1. Proposed expiry times for the [HTLC](#htlc)s.
```

>     1. Proposed expiry times on the for the [HTLC](#htlc)s. (Note: That could also be done by Bob.)

Author: After working in @D4nte 's comments the Note is gone. If that does not solve your issue please let me know.

> 1. Bob receives the swap proposed by Alice and sends back:

Reviewer 2: ```suggestion
1. Bob receives the swap proposed by Alice and sends back, if he agrees:
```

Author: We decided to remove that part :)

    1. His refund address on Bitcoin.
    1. His redeem address on Ethereum.

> 1. Alice has to fund first, she deploys an Ethereum HTLC transferring the agreed amount of ETH into the contract.

Reviewer 2: ```suggestion
1. Alice has to fund the swap first, she deploys an Ethereum HTLC transferring the agreed amount of ETH into the contract.
```

Author: Like it better the way it is.

> 1. Bob monitors the Ethereum ledger and notices that Alice has deployed the HTLC funding the ETH. 

Reviewer 2: ```suggestion
1. Bob monitors the Ethereum ledger and notices that Alice has deployed the HTLC funded it with the required amount of ether. 
```

He now sends a Bitcoin transaction into the Bitcoin network that is guarded by the script-hash of the Bitcoin HTLC (Bitcoin script).

> 1. Alice monitors the Bitcoin ledger and notices that Bob has sent the funding transaction. She can now redeem the Bitcoin by spending the Bitcoin by revealing the secret.

Reviewer 2: I had to think about this, and correct me if I'm wrong, but I think the asset 'bitcoin' would be spelled with a small 'b'.  The same way we write 'ten dollars' not 'ten Dollars'.  Same for ether.

Reviewer 2: 'spelled'/'spelt'  hmmm? - conjugating English verbs is a punish.

Reviewer 2: > This means that you should say "spelled" in US English and you can use both "spelt" and "spelled" in UK/Australian English. Spelled vs spelt: In American English, spelt primarily refers to the hardy wheat grown mostly in Europe, and the verb spell makes spelled in the past tense and as a past participle

1. Bob is monitoring the Bitcoin ledger and notices that Alice has spent from the funding transaction by revealing the secret.

> He extracts the actual secret from Alice's redeem transaction and uses it to redeem on Ethereum. 

Reviewer 2: ```suggestion
He extracts the secret from Alice's redeem transaction and uses it to redeem on Ethereum. 
```

<iframe 
    src="https://docs.google.com/presentation/d/e/2PACX-1vRjJeto_eQaZ8k5Ct00UzceHaSs-uIuvPmTFV6jP9SP6N-EKmct8H4cfN3xa2h-RFs8Gm4Gia41Je7m/embed?start=true&loop=false&delayms=1000" 
    frameborder="0" 
    width="801" 
    height="480" 
    allowfullscreen="true"
    mozallowfullscreen="true" 
    webkitallowfullscreen="true">
</iframe>

The COMIT protocol acts as enabler for atomic swaps in applications.
The [COMIT network daemon (cnd)](../comit-protocol/comit-protocol-stack.md#comit-network-daemon-cnd) mainly does two things:

1. Monitor the ledgers by processing blocks supplied by blockchain nodes.

> 1. Define the transaction details and hand them over to an application on top to send the actual signed transaction into the respective blockchain network.

Reviewer 2: Perhaps

```suggestion
1. Define the transaction details required for a swap and hand them over to an application built on top of COMIT.  (The application is then responsible for sending the actual [signed] transactions into the respective blockchain network.)
```

<iframe 
    src="https://docs.google.com/presentation/d/e/2PACX-1vTSToxZxKhOjfUWHVL5sjjjyGTdEXubjM3TpOpK-qR5Cjs7b6Tda9ZoX6n_NdM9iqgXBGOtPcPnjHAA/embed?start=false&loop=false&delayms=1000"
    frameborder="0"
    width="801"
    height="480"
    allowfullscreen="true"
    mozallowfullscreen="true"
    webkitallowfullscreen="true">
</iframe>

## Atomic Swap

> Atomic Swaps define a peer to peer exchange of two digital assets, where both parties are in control of their private keys throughout the whole execution of the swap.

Reviewer: I can see what you are saying here but it is not precisely true. If we were to add a wallet to COMIT, then we are still doing atomic swaps but the user is not in control of their funds :)

The main property of atomic swaps is atomicity, not the custodianship.

Author: I totally agree with what you say, but have problems of coming up with a precise definition. 
Can you, in your words, sum up what exactly the `atomicity` is? I tried and find it very hard.

Author: I can also just quote https://en.bitcoin.it/wiki/Atomic_swap if we are OK with that :)

Reviewer 2: I'll have a go, feel free to use it or not.

An Atomic Swap is an exchange of two digital assets between two parties in which both parties are guaranteed to either receive the other asset OR retain their original asset i.e., an atomic swap guarantees that neither party is able to take _both_ assets.

Reviewer: Here is my take:

An atomic swap allows two parties to exchange assets with certain atomicity guarantees. Within a given time window, the exchange is either completed (both parties receive the desired asset) or it is rolled back (both parties keep their original asset).

> The atomicity of the swap is achieved through using [HTLC](#htlc)s as a programmable escrow.

Reviewer: This reads like HTLCs are the _only_ way to do that.

Reviewer 2: Perhaps

In the case of COMIT, the atomicity of the swap is achieved ...

> Both parties require to have the same knowledge of all parameters for the HTLCs on both ledgers before they can start.

Reviewer 2: ```suggestion
Both parties are required to have the same knowledge of all parameters for the HTLCs on both ledgers before they can start.
```

Both parties then have the same functions available within the [HTLC](#htlc)s and can decide to move forward or wait to take their money back. 

> No party has access to both assets at the same time within the boundaries of the atomic swap. 

Reviewer: ```suggestion
No party has access to both assets at the same time within the time boundaries of the atomic swap. 
```

Note that the party in the cryptographic role of Alice has to redeem first, because she has the actual secret.

## HTLC

> HTLC stands for Hash Time Locked Contract. An HTLC is a script or smart contract that locks a digital asset over time.

Reviewer: Meta-question: Do we want to follow one sentence per line in these documents?

Author: yes, please correct if you find one that is not. sry for being a bit sloppy, tried to work relatively fast on that.

HTLCs require a ledger that provides a time lock.

<iframe 

>     src="https://docs.google.com/presentation/d/e/2PACX-1vRYb97VhvfyMa3oFC8CiVagBrNmOioSJpasERtSCi6RMwf0MwxCb1yTJBUm8_ZT9OOw6r-_fBsV23GX/embed?start=false&loop=false&delayms=3000"

Reviewer: This is pretty cool! :)

Author: Happy you like it, I hope google will not eventually decide to stop this :P

    frameborder="0"
    width="801"
    height="480"
    allowfullscreen="true"
    mozallowfullscreen="true"
    webkitallowfullscreen="true">
</iframe>

> HTLCs typically have three steps:

Reviewer: They have two steps but in the 2nd one you have the choice between two outcomes.

1. **Fund**: Locking up the asset in the HTLC.
1. **Redeem**: Until the expiry time-stamp is reached the other party has the change to redeem the locke-up assets with the secret.
1. **Refund**: After the expiry condition has been met, the party that locked up the asset can take it back.

In order to create HTLCs for the execution of an atomic swap, the two parties have to exchange certain parameters:

1. Ledgers, assets and amounts to be swapped.
1. Redeem and refund identity (address) of the respective parties.
1. Expiry times to be used for the HTLCs.
1. Hash of the secret (to be sent from the party in the cryptographic role of Alice to the party in the cryptographic role of Bob).

### HTLCs on different Ledgers

On Bitcoin this can be achieved by using a [Bitcoin script](https://en.bitcoin.it/wiki/Script).
The Bitcoin HTLC Script is defined in the [COMIT Bitcoin HTLC specification](https://github.com/comit-network/RFCs/blob/master/RFC-005-SWAP-Basic-Bitcoin.adoc): 

```

> OP_IF

Reviewer: Can we do this inlining automatically with docusaurus?

Author: you mean inlining the code from github? 

If this it what you mean: I played around a bit using gists and stuff, but could not really make it work.

    OP_SIZE 32 OP_EQUALVERIFY
    OP_SHA256 <secret_hash> OP_EQUALVERIFY
    OP_DUP OP_HASH160 <redeem_identity>
OP_ELSE
    <refund_timestamp> OP_CHECKLOCKTIMEVERIFY OP_DROP
    OP_DUP OP_HASH160 <refund_identity>
OP_ENDIF
OP_EQUALVERIFY
OP_CHECKSIG
```

On Ethereum this can be achieved by deploying an [Ethereum Smart Contract](https://solidity.readthedocs.io/en/v0.6.2/introduction-to-smart-contracts.html).
Note that an Ethereum HTLC does not necessarily have to be written in Solidity.
Given that HTLCs are really simple smart contracts, they can also be written in EVM assembly as defined in the [COMIT Ether HTLC specification](https://github.com/comit-network/RFCs/blob/master/RFC-007-SWAP-Basic-Ether.adoc) and [COMIT ERC20 HTLC specification](https://github.com/comit-network/RFCs/blob/master/RFC-009-SWAP-Basic-ERC20.adoc).

[Scriptless Scripts](privacy-preserving-swap.md#scriptless-scripts) enable HTLCs hidden in signatures. This allows [Privacy Preserving Swaps](privacy-preserving-swap.md).

### Close Look on HTLC Expiries

The expiries of the HTLCs have to be defined so that Alice's expiry time is further in the future than Bob's.

> This is because Alice is protected by the secret, and if she choses not to reveal the secret, Bob must have the change to refund early.

Reviewer 2: ```suggestion
This is because Alice is protected by the secret, and if she chooses not to reveal the secret, Bob must have the chance to refund early.
```

Choosing the right initial expiries, and defining if is is still *safe* for a party to move forward during swap execution is not trivial.

> The presentation slides below show some scenarios to demonstrate the motivation of both parties to act according to the protocol during swap execution.

Reviewer 2: ```suggestion
The presentation slides below shows some scenarios to demonstrate the motivation of both parties to act according to the protocol during swap execution.
```

---
id: atomic-swap-htlc
title: Atomic Swaps using HTLCs
sidebar_label: Atomic Swaps using HTLCs
---

> This documentation section provides an introduction to atomic swaps, hash time locked contracts (HTLCs) and how these concepts are used in COMIT.

Author: Given, that we write `Hash Time Locked Contract` everywhere else, I would go with that. And yes, it should be uniform. Discussed this with Tobin as well, and he said he would deem `Hash Time Locked Contract` the correct way to write it.

Atomic Swaps and HTLCs are core concepts of the COMIT protocol.

> <iframe 

Reviewer 3: I think this whole slide should be after the 'Atomic Swap' and 'HTLC' subsections, since readers wouldn't know what the terms on the slide (e.g. HTLC) means before these subsections are read.

Author: The slide is a good attention grabber, that's why we put it there. It is not supposed to explain everything, but motivate the user to have a closer look on the content below.

I strongly vote for leaving it at the top, knowing that it does not answer all questions - other opinions? @bonomat @thomaseizinger

    src="https://docs.google.com/presentation/d/e/2PACX-1vRjJeto_eQaZ8k5Ct00UzceHaSs-uIuvPmTFV6jP9SP6N-EKmct8H4cfN3xa2h-RFs8Gm4Gia41Je7m/embed?start=true&loop=false&delayms=1000" 
    frameborder="0" 
    width="801" 
    height="480" 
    allowfullscreen="true"
    mozallowfullscreen="true" 
    webkitallowfullscreen="true">
</iframe>

This slide shows an overview of how atomics swaps using HTLCs work in COMIT, more details can be found in the sections below.

## Atomic swap

An Atomic Swap is an exchange of two digital assets between two parties in which both parties are guaranteed to either receive the other asset OR retain their original asset i.e., an atomic swap guarantees that neither party is able to take both assets.

> Atomicity of the swap can be achieved by [HTLC](#hash-time-locked-contract-htlc)s that serve as a programmable escrow, or other concepts as explained in the [Bitcoin Wiki](https://en.bitcoin.it/wiki/Atomic_swap).

Reviewer 3: I don't think we should tell people to go elsewhere for an explanation. At the very least, what do you think of putting some of the explanation here?

Author: Since `core-concepts` is about explaining things in a more general way, I would actually say we **should** link to other documentation. The Bitcoin Wiki is written in a nice way and I see no value in rephrasing that. I don't think the link hurts us, on the contrary. (The Bitcoin Wiki is also cited by other documentation like https://liquality.io/blog/hash-time-locked-contracts-htlcs-explained/) - other opinions?

Reviewer 3: One alternative is to add a brief snippet of the most important part from the wiki.

## Hash Time Locked Contract (HTLC)

HTLC stands for Hash Time Locked Contract. An HTLC is a script or smart contract that locks a digital asset over time.
HTLCs require a ledger that provides a time lock.

> Both parties are required to have the same knowledge of all parameters for the HTLCs on both ledgers before they can start.

Reviewer 3: What are some examples of these parameters?

Reviewer 3: It doesn't have to be tied strictly to our implementation, but knowing a high-level interface would be helpful in understanding the concepts I think.

> Both parties then have the same functions available within the HTLCs and can decide to move forward or wait to take their money back. 

Reviewer 3: What are the functions that parties have access to?

> No party has access to both assets at the same time within the time boundaries of the atomic swap. 

Reviewer 3: Why does this matter?

HTLCs typically define three functions:

1. **Fund**: Locking up the asset in the HTLC.
1. **Redeem**: Until the expiry time-stamp is reached the other party has the chance to take the locked-up assets with the secret.
1. **Refund**: After the expiry condition has been met, the party that locked up the asset can take it back.

> HTLCs are secured by a time-lock and hash-lock.

Reviewer 3: There are no mention of the time-lock and how it's used after this sentence, only the hash-lock is explained.

> The hash-lock requires the generation of a secret prior to funding. 

Reviewer 3: Could we split out this bit on the hash-lock to a separate paragraph, or as one of two bullet points?

> The order of the actions is then tied to the cryptographic roles of Alice Bob. 

Reviewer 3: 'Alice Bob' -> 'Alice and Bob'

> Alice is the role that comes up with the secret.

Reviewer 3: This multi-step process may be better explained by referencing a diagram, perhaps a screenshot of one of the slides.

She shares the hash of that secret with Bob enabling him to fund on his side.
The party in the cryptographic role of Alice has to fund and redeem first, because she has the actual secret.

> ## Atomic Swaps using HTLCs in COMIT

Reviewer 3: I think this section should be relabelled as 'Negotiation' instead.

Reviewer 3: The original heading can be assigned to the paragraphs after it.

Author: It seems that the first sentence of this section, mentioning `Negotiation` is confusing. It can be removed. I put it there to have a link to the negotiation phase to try to create a bit more context.

> Prior to the actual swap execution there is usually a [Negotiation](negotiation.md) phase where the two parties agree on the ledgers, assets and amounts to be swapped.

Reviewer 3: What do you think of having a diagram on how the negotiation fit into the larger swap process?

Author: We should have that, but not in this section. There is a separate `Negotiation` file describing this.

Reviewer 3: I see. At this point of the document, the question that comes to mind is 'How does the Negotiation fit into the bigger picture?'

> Assuming that Alice has ETH on Ethereum and Bob has BTC on Bitcoin, Alice wants Bob's BTC and Bob wants Alice's ETH, the execution of a swap would be like this:

Reviewer 3: I feel that these steps should be better integrated with the visuals we already have. What do you think?

1. Alice generates a secret and hashes it. She sends to Bob:
    1. Hash of the secret. 
    1. Her refund address on Ethereum.
    1. Her redeem address on Bitcoin.
    1. Proposed expiry times for the HTLCs.
1. Bob receives the swap proposed by Alice and sends back:
    1. His refund address on Bitcoin.
    1. His redeem address on Ethereum.
1. Alice has to fund first, she deploys an Ethereum HTLC transferring the agreed amount of ETH into the contract.
1. Bob monitors the Ethereum ledger and notices that Alice has deployed the HTLC funding the ETH. 
He now sends a Bitcoin transaction into the Bitcoin network that is guarded by the script-hash of the Bitcoin HTLC (Bitcoin script).
1. Alice monitors the Bitcoin ledger and notices that Bob has sent the funding transaction. She can now redeem the Bitcoin by spending the Bitcoin by revealing the secret.
1. Bob is monitoring the Bitcoin ledger and notices that Alice has spent from the funding transaction by revealing the secret.
He extracts the actual secret from Alice's redeem transaction and uses it to redeem on Ethereum. 

> The COMIT protocol acts as enabler for atomic swaps in applications.

Reviewer 3: This could have its own section heading e.g. 'Atomic Swaps using HTLCs in COMIT'.

The [COMIT network daemon (cnd)](../comit-protocol/comit-protocol-stack.md#comit-network-daemon-cnd) mainly does two things:

1. Monitor the ledgers by processing blocks supplied by blockchain nodes.

> 1. Define the transaction details required for a swap and hand them over to an application built on top of COMIT.  The application is then responsible for sending the actual signed transactions into the respective blockchain network.

Author: worked in with Thomas comments, combined them in suggestion

<iframe 
    src="https://docs.google.com/presentation/d/e/2PACX-1vTSToxZxKhOjfUWHVL5sjjjyGTdEXubjM3TpOpK-qR5Cjs7b6Tda9ZoX6n_NdM9iqgXBGOtPcPnjHAA/embed?start=false&loop=false&delayms=1000"
    frameborder="0"
    width="801"
    height="480"
    allowfullscreen="true"
    mozallowfullscreen="true"
    webkitallowfullscreen="true">
</iframe>

### HTLCs in COMIT

> In order to create HTLCs for the execution of an atomic swap, the two parties have to exchange certain parameters:

Reviewer 3: 'certain' -> 'the following'

1. Ledgers, assets and amounts to be swapped.
1. Redeem and refund identity (address) of the respective parties.
1. Expiry times to be used for the HTLCs.

> 1. Hash of the secret (to be sent from the party in the cryptographic role of Alice to the party in the cryptographic role of Bob).

Reviewer 3: Might be better to split the details into a separate sentence or removed entirely:

```
Hash of the secret. This hash is sent from the party in the cryptographic role of Alice to the party in the cryptographic role of Bob.
```

<iframe 
    src="https://docs.google.com/presentation/d/e/2PACX-1vRYb97VhvfyMa3oFC8CiVagBrNmOioSJpasERtSCi6RMwf0MwxCb1yTJBUm8_ZT9OOw6r-_fBsV23GX/embed?start=false&loop=false&delayms=3000"
    frameborder="0"
    width="801"
    height="480"
    allowfullscreen="true"
    mozallowfullscreen="true"
    webkitallowfullscreen="true">
</iframe>

> ### HTLCs on different Ledgers

Reviewer 3: How about splitting into:
- HTLCs on Bitcoin
- HTLCs on Ethereum

The Bitcoin HTLC Script is defined in the [COMIT Bitcoin HTLC specification](https://github.com/comit-network/RFCs/blob/master/RFC-005-SWAP-Basic-Bitcoin.adoc): 

```
OP_IF
    OP_SIZE 32 OP_EQUALVERIFY
    OP_SHA256 <secret_hash> OP_EQUALVERIFY
    OP_DUP OP_HASH160 <redeem_identity>
OP_ELSE
    <refund_timestamp> OP_CHECKLOCKTIMEVERIFY OP_DROP
    OP_DUP OP_HASH160 <refund_identity>
OP_ENDIF
OP_EQUALVERIFY
OP_CHECKSIG
```

On Ethereum this can be achieved by deploying an [Ethereum Smart Contract](https://solidity.readthedocs.io/en/v0.6.2/introduction-to-smart-contracts.html).
Note that an Ethereum HTLC does not necessarily have to be written in Solidity.
Given that HTLCs are really simple smart contracts, they can also be written in EVM assembly as defined in the [COMIT Ether HTLC specification](https://github.com/comit-network/RFCs/blob/master/RFC-007-SWAP-Basic-Ether.adoc) and [COMIT ERC20 HTLC specification](https://github.com/comit-network/RFCs/blob/master/RFC-009-SWAP-Basic-ERC20.adoc).

[Scriptless Scripts](privacy-preserving-swap.md#scriptless-scripts) enable HTLCs hidden in signatures. This allows [Privacy Preserving Swaps](privacy-preserving-swap.md).

> ### HTLC Expiry Details

Reviewer 3: I think you can remove 'Details' from 'HTLC Expiry Details.'

Atomic Swaps and HTLCs are core concepts of the COMIT protocol.

<iframe 
    src="https://docs.google.com/presentation/d/e/2PACX-1vRjJeto_eQaZ8k5Ct00UzceHaSs-uIuvPmTFV6jP9SP6N-EKmct8H4cfN3xa2h-RFs8Gm4Gia41Je7m/embed?start=true&loop=false&delayms=1000" 
    frameborder="0" 
    width="801" 
    height="480" 
    allowfullscreen="true"
    mozallowfullscreen="true" 
    webkitallowfullscreen="true">
</iframe>

This slide shows an overview of how atomics swaps using HTLCs work in COMIT, more details can be found in the sections below.

## Atomic swap

An Atomic Swap is an exchange of two digital assets between two parties in which both parties are guaranteed to either receive the other asset OR retain their original asset i.e., an atomic swap guarantees that neither party is able to take both assets.

Atomicity of the swap can be achieved by [HTLC](#hash-time-locked-contract-htlc)s that serve as a programmable escrow, or other concepts as explained in the [Bitcoin Wiki](https://en.bitcoin.it/wiki/Atomic_swap).

## Hash Time Locked Contract (HTLC)

> HTLC stands for Hash Time Locked Contract. An HTLC is a script or smart contract that locks a digital asset over time.

Reviewer: ```suggestion
HTLC stands for Hash Time Locked Contract.
An HTLC is a script or smart contract that locks a digital asset over time.
```

> HTLCs require a ledger that provides a time lock.

Reviewer: ```suggestion
HTLCs only work on ledgers with time locks.
```

1. Bob is monitoring the Bitcoin ledger and notices that Alice has spent from the funding transaction by revealing the secret.
He extracts the actual secret from Alice's redeem transaction and uses it to redeem on Ethereum. 

The COMIT protocol acts as enabler for atomic swaps in applications.
The [COMIT network daemon (cnd)](../comit-protocol/comit-protocol-stack.md#comit-network-daemon-cnd) mainly does two things:

1. Monitor the ledgers by processing blocks supplied by blockchain nodes.

> 1. Define the transaction details required for a swap and hand them over to an application built on top of COMIT.  The application is then responsible for sending the actual signed transactions into the respective blockchain network.

Reviewer: ```suggestion
1. Define the transaction details required for a swap and hand them over to an application built on top of COMIT.
The application is then responsible for sending the actual signed transactions into the respective blockchain network.
```

    webkitallowfullscreen="true">
</iframe>

### HTLCs on different Ledgers

> On Bitcoin HTLCs can be achieved by using a [Bitcoin script](https://en.bitcoin.it/wiki/Script).

Reviewer: ```suggestion
On Bitcoin, HTLCs can be achieved by using a [Bitcoin script](https://en.bitcoin.it/wiki/Script).
```
